### PR TITLE
AI admin settings

### PR DESCRIPTION
**Pull request by marcelklehr**
* Resolves: -
* Follow up on #38854 

## Summary
Adds an admin settings section for AI settings, allowing admins to specify the precedence of machine translation providers, set the speech-to-text provider to use and select which text processing provider to use for which task.

image

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by susnux** [comment]
Nice! But keep in mind vuedragable is not accessible, so there is no way to reorder by using the keyboard, this needs to be done by the user of the component itself.

**Comment by marcelklehr** [comment]
@susnux How would you go about adding a11y here?

**Comment by susnux** [comment]
> How would you go about adding a11y here?

For Forms I suggested this one: https://github.com/nextcloud/forms/pull/1532

So basically either implement key event listeners (arrow up / down) or add additional buttons for ordering

**Comment by marcelklehr** [comment]
image

File: apps/settings/src/components/AdminAI.vue [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+					:value="provider.class"
+					name="stt_provider"
+					type="radio"
+					@update:checked="saveChanges">
+					{{ provider.name }}
+				</NcCheckboxRadioSwitch>
+			</template>
+			<template v-if="sttProviders.length === 0">
+				<NcCheckboxRadioSwitch disabled type="radio">
+					{{ t('settings', 'No apps are currently installed that provide Speech-To-Text functionality') }}
```
Comment (julien-nc) [inline apps/settings/src/components/AdminAI.vue:25]:
```suggestion
					{{ t('settings', 'None of your currently installed apps provide Speech-To-Text functionality') }}
```

File: apps/settings/src/components/AdminAI.vue [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+						</template>
+						<template #selected-option="{label}">
+							{{ textProcessingProviders.find(p => p.class === label)?.name }}
+						</template>
+					</NcSelect>
+					<p>&nbsp;</p>
+				</div>
+			</template>
+			<template v-if="Object.keys(settings['ai.textprocessing_provider_preferences']).length === 0 || !Array.isArray(textProcessingTaskTypes)">
+				<p>{{ t('settings', 'No apps are currently installed that provide Text processing functionality') }}</p>
```
Comment (julien-nc) [inline apps/settings/src/components/AdminAI.vue:51]:
```suggestion
				<p>{{ t('settings', 'None of your currently installed apps provide Text processing functionality') }}</p>
```

File: apps/settings/src/components/PersonalInfo/AvatarSection.vue [hunk 3]
```diff
@@ -22,21 +22,23 @@
 
 <template>
 	<section id="vue-avatar-section">
-		<h3 class="hidden-visually"> {{ t('settings', 'Your profile information') }} </h3>
+		<h3 class="hidden-visually">
+			{{ t('settings', 'Your profile information') }}
+		</h3>
```
Comment (julien-nc) [inline apps/settings/src/components/PersonalInfo/AvatarSection.vue:27]: A bit out of scope but why not. :grin:

**Review by julien-nc — CHANGES_REQUESTED** [review]
Awesome, tested with a FreePrompt, a Summarize and an Outline TP provider.
The fallback works just fine if a provider fails.

Small string changes requested but all the rest LGTM.

I'm also wondering if all the unrelated stuff belongs here but since these are nice fixes, why not letting them in this.

About accessibility, there could be 2 buttons to move items up and down.

File: apps/settings/src/components/PersonalInfo/AvatarSection.vue (hunk 3 above)
Comment (marcelklehr) [reply apps/settings/src/components/PersonalInfo/AvatarSection.vue:27]: Yeah, I added one commit that does a lint:fix :shrug:

**Comment by marcelklehr** [comment]
> About accessibility, there could be 2 buttons to move items up and down.

Already added :) (see screenshot above)

File: apps/settings/src/components/AdminAI.vue [hunk 4]
```diff
@@ -0,0 +1,141 @@
+<template>
+	<div>
+		<NcSettingsSection :title="t('settings', 'Machine translation')"
+			:description="t('settings', 'Machine translation can be implemented by different apps. Here you can define the precedence of the machine translation apps you have installed at the moment.')">
+			<draggable v-model="settings['ai.translation_provider_preferences']" @change="saveChanges">
+				<div v-for="(providerClass, i) in settings['ai.translation_provider_preferences']" :key="providerClass" class="draggable__item">
+					<DragVerticalIcon /> <span class="draggable__number">{{ i+1 }}</span> {{ translationProviders.find(p => p.class === providerClass)?.name }}
+					<NcButton aria-label="Move up" type="tertiary" @click="settings['ai.translation_provider_preferences'].splice(Math.min(i-1,0), 0, ...settings['ai.translation_provider_preferences'].splice(i, 1)); saveChanges()"><template #icon><ArrowUpIcon /></template></NcButton>
+					<NcButton aria-label="Move down" type="tertiary" @click="settings['ai.translation_provider_preferences'].splice(i+1, 0, ...settings['ai.translation_provider_preferences'].splice(i, 1)); saveChanges()"><template #icon><ArrowDownIcon /></template></NcButton>
```
Comment (Pytal) [inline apps/settings/src/components/AdminAI.vue:9]: Should move the long click handlers into methods

File: apps/settings/src/components/PersonalInfo/AvatarSection.vue (hunk 3 above)
Comment (Pytal) [reply apps/settings/src/components/PersonalInfo/AvatarSection.vue:27]: Thanks for the lint fixes! Let's keep these scoped next time as there are a lot 🤯

File: core/Migrations/Version28000Date20230728104802.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+				'notnull' => true,
+				'length' => 32,
+				'default' => '',
+			]);
+			$table->addColumn('identifier', Types::STRING, [
+				'notnull' => true,
+				'length' => 255,
+				'default' => '',
+			]);
+			$table->addColumn('last_updated', 'integer', [
```
Comment (Pytal) [inline core/Migrations/Version28000Date20230728104802.php:91]:
```suggestion
			$table->addColumn('last_updated', Types::INTEGER, [
```

**Review by julien-nc — APPROVED** [review]

**Comment by julien-nc** [comment]
@Pytal Thanks for the review. Your changes have been implemented.
Rebased on master.

**Review by Pytal — APPROVED** [review]

**Comment by julien-nc** [comment]
Cypress failure is related with user settings.
acceptance-app-files failure is related with a file action menu button.

**Comment by kesselb** [comment]
The failing TextProcessingTest looks related:

image

**Comment by kesselb** [comment]
Fix: https://github.com/nextcloud/server/pull/39684

**Pull request merged**